### ISSUE TITLE
Compacting interrupts reading/writing?

### ISSUE THREAD

**Issue by ettom** [njaard/sonnerie#31]
Not sure if I'm misunderstanding something, but in the README it says
>Compacting doesn't block readers or writers, but only one can happen at any given moment, so a lock is placed to prevent multiple concurrent compactions.
>Compactions are atomic, so you can cancel it (with ^C) at any time.

When doing
```sh
mkdir db
touch db/main
```
And then, simultenously running
```
while : ; do sonnerie -d db compact; done
```
and
```
for i in $(seq 1 10000); do echo $i; echo "test $i $i" | sonnerie -d db add --format u; done
```

The writing occasionally panics:
`thread 'main' panicked at 'opening db: Os { code: 2, kind: NotFound, message: "No such file or directory" }'` and compacting gives the following message `disregarding "db/tx.17a24759f7b343cb", it is zero length`. Whenever writing panics, the number for that iteration is missing from the database.

Is this to be expected? Obviously it's an unrealistic scenario, but as far as I can tell, this means there is a non-0 chance this can happen during normal operation too.

**Comment by ettom** [comment]
Similarly, doing 
```
while :; do sonnerie -d db read % || break; done
```
on a database with about a 1000 rows and then running
```
sonnerie -d db compact
```
will result in the loop stopping with `Error: Os { code: 2, kind: NotFound, message: "No such file or directory" }`.

**Comment by njaard** [comment]
What operating system are you on?

**Comment by ettom** [comment]
Arch Linux. I also tried this on a custom LFS-like embedded Linux system and got the same results. sonnerie v0.8.5 and ext4 for both.

**Comment by ettom** [comment]
Interesting, this doesn't seem to happen on a Linux Mint 21 machine with ext4. On that machine I the compaction loop just says `disregarding "db/tx.*", it is zero length` on every iteration.

**Comment by ettom** [comment]
> Interesting, this doesn't seem to happen on a Linux Mint 21 machine with ext4. On that machine the compaction loop just says `disregarding "db/tx.*", it is zero length` on every iteration.

Scratch that, it still happens, but a *lot* less often. Easy to catch it with
```
for i in $(seq 1 100000); do echo $i; echo "test $i $i" | sonnerie -d db add --format u || break; done
```

**Comment by njaard** [comment]
It seems like there's a race condition in the time between the completion of [`read_dir`](https://github.com/njaard/sonnerie/blob/master/src/database_reader.rs#L74) and the [actual opening of the files](https://github.com/njaard/sonnerie/blob/master/src/database_reader.rs#L106). If a compaction completes in between those two actions, then some files that were previously found via the `read_dir` will have been deleted by the compaction.

If that happens, then I think the correct action is to try the `read_dir` again from the beginning. You can't simply ignore the missing files, because your `read_dir` necessarily will not have opened the file they'll have been compacted into.

**Comment by njaard** [comment]
Fixed in commit [322e423](https://github.com/njaard/sonnerie/commit/322e423d777a2e862bb46661fae748a356b2fdbe)
